### PR TITLE
Adding language support for JS/TS via typescript-language-server

### PR DESCRIPTION
**Pull request by themichaelusa**
This is still a work in progress, but should be ready to go ASAP.

**TODOs:**
- [x] Revert changes to `pyproject.toml`
- [x] Remove junk files from my fork (`todo.txt`,` tests_js`, etc)
- [x] Proper tests for JS/TS integration under `multilspy/tests`
- [x] Some code styling nits + comments where needed
- [x] Revert event loop changes  (these are for a separate PR - assuming I can replicate hanging behavior w/ SyncLanguageServer)

### CODE REVIEW

**Comment by themichaelusa** [comment]
@microsoft-github-policy-service agree [company="Rubrick, Inc."]

**Comment by themichaelusa** [comment]
@microsoft-github-policy-service agree company="Rubrick, Inc."

**Comment by mrT23** [comment]
I think i solved the problem described here:
https://github.com/microsoft/multilspy/issues/5#issuecomment-2405189497

see attached file:
[initialize_params.json](https://github.com/user-attachments/files/17373119/initialize_params.json)

I basically removed the `window` section from the `capabilities`. I don't think it is needed for our API usage, and it causes the stuck problem

image

**Comment by themichaelusa** [comment]
> I think i solved the problem described here: [#5 (comment)](https://github.com/microsoft/multilspy/issues/5#issuecomment-2405189497)
> 
> see attached file: [initialize_params.json](https://github.com/user-attachments/files/17373119/initialize_params.json)
> 
> I basically removed the `window` section from the `capabilities`. I don't think it is needed for our API usage, and it causes the stuck problem
> 
> image

Ok great. Had a busy weekend -  will write test cases tonight and we should be ready to merge.

**Comment by LakshyAAAgrawal** [comment]
Dear @themichaelusa  and @mrT23 ,

Thank you very much for your efforts on implementing JS/TS support. I will try to review this ASAP.

Excited about this development :)

File: src/multilspy/language_servers/typescript_language_server/typescript_language_server.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            await self.server.start()
+            initialize_params = self._get_initialize_params(self.repository_root_path)
+
+            self.logger.log(
+                "Sending initialize request from LSP client to LSP server and awaiting response",
+                logging.INFO,
+            )
+            init_response = await self.server.send.initialize(initialize_params)
+            
+            # # TypeScript-specific capability checks
```
Comment (LakshyAAAgrawal) [inline src/multilspy/language_servers/typescript_language_server/typescript_language_server.py:168]: Please try to include as many of these checks as possible, and uncomment them.

File: tests/multilspy/test_multilspy_javascript.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+"""
+This file contains tests for running the Python Language Server: jedi-language-server
```
Comment (LakshyAAAgrawal) [inline tests/multilspy/test_multilspy_javascript.py:2]: Can you update the module level comment to reflect js?

File: tests/multilspy/test_multilspy_typescript.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+"""
+This file contains tests for running the Python Language Server: jedi-language-server
```
Comment (LakshyAAAgrawal) [inline tests/multilspy/test_multilspy_typescript.py:2]: nit: Module level comment -> typescript

File: tests/multilspy/test_sync_multilspy_javascript.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
+"""
+This file contains tests for running the Python Language Server: jedi-language-server
```
Comment (LakshyAAAgrawal) [inline tests/multilspy/test_sync_multilspy_javascript.py:2]: nit: Module level comment to include javascript

File: tests/multilspy/test_sync_multilspy_typescript.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+"""
+This file contains tests for running the Python Language Server: jedi-language-server
```
Comment (LakshyAAAgrawal) [inline tests/multilspy/test_sync_multilspy_typescript.py:2]: nit: Module level comment to reflect typescript

**Review by LakshyAAAgrawal — COMMENTED** [review]
Just went over the code. Looks great. Will try to test it on my system soon and then merge. In the meanwhile, could you please address the few nits I have added?

File: pyproject.toml
[the remark this answers is not in the document]
Comment (themichaelusa) [reply pyproject.toml:1]: I diffed this version of the `pyproject.toml` with the original one that you committed 2 months ago and they seem to be identical w/ no whitespace changes... might just be Github erroneously flagging it?

File: src/multilspy/language_servers/typescript_language_server/typescript_language_server.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        super().__init__(
+            config,
+            logger,
+            repository_root_path,
+            ProcessLaunchInfo(cmd=ts_lsp_executable_path, cwd=repository_root_path),
+            "typescript",
+        )
+        self.server_ready = asyncio.Event()
+
+    def setup_runtime_dependencies(self, logger: MultilspyLogger, config: MultilspyConfig) -> str:
```
[the remark this answers is not in the document]
Comment (themichaelusa) [reply src/multilspy/language_servers/typescript_language_server/typescript_language_server.py:43]: Am pretty confident that all the dependencies required for the LSP to work will fall under the `src/multilspy/language_servers/typescript_language_server/static/node_modules` directory, as we aren't using the `npm -g` flag when we are installing the Typescript / LSP packages. 

^ I verified this bit ago, but feel free to take another look!

The only external dependency we expect is a node runtime, but we expect the same for Python, Java, etc.

File: src/multilspy/language_servers/typescript_language_server/typescript_language_server.py (hunk 1 above)
Comment (themichaelusa) [reply src/multilspy/language_servers/typescript_language_server/typescript_language_server.py]: Done!

**Comment by themichaelusa** [comment]
> Just went over the code. Looks great. Will try to test it on my system soon and then merge. In the meanwhile, could you please address the few nits I have added?

Done! Please let me know if there are any issues running it on your end.

 Portability nit: I don't have the means to test Windows support, but I've been testing this on MacOS locally, and we've been running this fork (`prod` branch) in production on GCP - so we can be reasonably confident re: Linux support.

File: tests/multilspy/test_multilspy_typescript.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            for item in result:
+                del item["uri"]
+                del item["absolutePath"]
+
+            assert result == [
+                {'range': {'start': {'line': 231, 'character': 15}, 'end': {'line': 231, 'character': 21}}, 'relativePath': path}, 
+                {'range': {'start': {'line': 264, 'character': 12}, 'end': {'line': 264, 'character': 18}}, 'relativePath': path}
+            ]
+
+            print(result)
```
Comment (LakshyAAAgrawal) [inline tests/multilspy/test_multilspy_typescript.py:56]: Can you remove this print line?

**Review by LakshyAAAgrawal — CHANGES_REQUESTED** [review]
I have reviewed the full PR and verified it running correctly. Thank you very much for this PR.

Can you make this one final change (remove print statement), and I will immediately proceed to merge.

File: tests/multilspy/test_multilspy_typescript.py (hunk 7 above)
Comment (themichaelusa) [reply tests/multilspy/test_multilspy_typescript.py]: Done!

**Comment by themichaelusa** [comment]
> I have reviewed the full PR and verified it running correctly. Thank you very much for this PR.
> 
> Can you make this one final change (remove print statement), and I will immediately proceed to merge.

No problem! I look forward to opening many future PRs. This is a great package and a has a lot of potential

**Review by LakshyAAAgrawal — APPROVED** [review]

**Comment by shobrook** [comment]
Love to see this

**Pull request merged**